rust/src/ffi/evmc_vm.rs

```diff
@@ -0,0 +1,211 @@
+    instance: *mut evmc_vm_t,
+    key: *const c_char,
+    value: *const c_char,
+) -> evmc_set_option_result {
+    assert!(!instance.is_null());
+
+    if key.is_null() {
+        return evmc_set_option_result::EVMC_SET_OPTION_INVALID_NAME;
+    }
+
+    let key = unsafe { CStr::from_ptr(key) };
+    let Ok(key) = key.to_str() else {
+        return evmc_set_option_result::EVMC_SET_OPTION_INVALID_NAME;
+    };
+
+    let value = if !value.is_null() {
+        unsafe { CStr::from_ptr(value) }
+    } else {
+        unsafe { CStr::from_bytes_with_nul_unchecked(&[0]) }
+    };
+
+    let Ok(value) = value.to_str() else {
+        return evmc_set_option_result::EVMC_SET_OPTION_INVALID_VALUE;
+    };
+
+    let mut container = unsafe {
+        // Acquire ownership from EVMC.
+        EvmcContainer::<EvmRs>::from_ffi_pointer(instance)
+    };
+
+    let result = match container.set_option(key, value) {
+        Ok(()) => evmc_set_option_result::EVMC_SET_OPTION_SUCCESS,
+        Err(SetOptionError::InvalidKey) => evmc_set_option_result::EVMC_SET_OPTION_INVALID_NAME,
+        Err(SetOptionError::InvalidValue) => evmc_set_option_result::EVMC_SET_OPTION_INVALID_VALUE,
+    };
+
+    // Release ownership to EVMC.
+    unsafe {
+        EvmcContainer::into_ffi_pointer(container);
+    }
+
+    result
+}
+
+#[no_mangle]
+pub extern "C" fn evmc_create_evmrs() -> *const evmc_vm_t {
+    let new_instance = evmc_vm_t {
+        abi_version: EVMC_ABI_VERSION as i32,
+        destroy: Some(__evmc_destroy),
+        execute: Some(__evmc_execute),
+        get_capabilities: Some(__evmc_get_capabilities),
+        set_option: Some(__evmc_set_option),
+        name: unsafe { CStr::from_bytes_with_nul_unchecked(EVM_RS_NAME.as_bytes()).as_ptr() },
+        version: unsafe { CStr::from_bytes_with_nul_unchecked(EVM_RS_VERSION.as_bytes()).as_ptr() },
+    };
+
+    let container = EvmcContainer::<EvmRs>::new(new_instance);
+
+    // Release ownership to EVMC.
+    unsafe { EvmcContainer::into_ffi_pointer(container) }
+}
+
+extern "C" fn __evmc_destroy(instance: *mut evmc_vm_t) {
+    if instance.is_null() {
+        // This is an irrecoverable error that violates the EVMC spec.
+        process::abort();
+    }
+    unsafe {
+        // Acquire ownership from EVMC. This will deallocate it also at the end of the scope.
+        EvmcContainer::<EvmRs>::from_ffi_pointer(instance);
+    }
+}
+
+extern "C" fn __evmc_execute(
+    instance: *mut evmc_vm_t,
+    host: *const evmc_host_interface,
+    context: *mut evmc_host_context,
+    revision: evmc_revision,
+    message: *const evmc_message,
+    code: *const u8,
+    code_size: usize,
+) -> evmc_result {
+    if instance.is_null()
+        || (host.is_null() && EVMC_CAPABILITY != evmc_capabilities::EVMC_CAPABILITY_PRECOMPILES)
+        || message.is_null()
+        || (code.is_null() && code_size != 0)
+    {
+        // These are irrecoverable errors that violate the EVMC spec.
+        std::process::abort();
```

Comment:
I am not sure about this.
On the one hand, I understand the validity of rejecting ill-formed context. On the other hand, I see two problems with this: 

- The conditions checked in this have different criticality:
   - this code is indeed odd, but see second paragraph about abort. 
   - capability is an explicit check: that should be done here
   - null instance could be the sign of an ill-formed program. See second paragraph about this.
   - null host could be ignored until callbacks are needed, but it may be better to check it here nevertheless. Execution error could be returned. 
   - I am not sure about message. I am not sure what does it contain, and a default value could mask the error completely and return a successful execution. 

- We have no guarantee at all about the correctness or nature of the environment calling this library. In production would be pretty bad if this crashes, and we have no means to test if every input will be always valid.  In other words: murphy's law, if you do not write an abort here, it cannot abort here. 
    - Would it be fundamentally wrong to return an error? EVMC_FAILURE or EVMC_INTERNAL_ERROR ? 

I see that the abort comes from the original implementation: https://docs.rs/evmc-declare/latest/src/evmc_declare/lib.rs.html#350
Nevertheless, It may be sensible to avoid killing the process.

Reply:
I understand that you would like to avoid aborting. If there are sensible defaults for ill formed values that could be used instead, I would be fine with that. And if we can return an error that would be even better but here I'm unsure about the semantics. However I would like to avoid deferring check until first use because then we have those checks all over the place and not (like currently) in a single place.

Reply:
Unless @HerbertJordan has a definite opinion on what to do here we could also just open an issue to not forget about it and merge it as is. This way all the other stuff can get merged more quickly and changing this later on does not influence the other parts of the code anyway.

Reply:
If you are unsure about the semantics, we leave as it is. After all, this is a copy of the original macro expansion, and not a redo.  
The one option that we could explore in the future, is what happens if we return an error here instead of abort.